### ISSUE TITLE
chainsync decode MsgRollForwardNtC error

### ISSUE THREAD

**Issue by zhp1254** [blinklabs-io/gouroboros#197]
protocol/chainsync/client.go
line : 210
blk, err := ledger.NewBlockFromCbor(msg.BlockType(), msg.BlockCbor())

but BlockType and BlockCbor did not give a value in
protocol/chainsync/messages.go
line : 45
ret = &MsgRollForwardNtC{}

**Comment by agaffney** [comment]
A little more context would be helpful. How are you using the library? What network was this? What slot/block?

**Comment by zhp1254** [comment]
cmd:
./go-ouroboros-network  -network mainnet -socket /cardano-node/state-node-mainnet/node.socket chain-sync

tag:
v0.29.0

**Comment by zhp1254** [comment]
./go-ouroboros-network -network mainnet -ntn -address 127.0.0.1:3001 chain-sync -tip
{"MessageType":6,"Tip":{"Point":{"Slot":85162758,"Hash":"6kV9P2zOMOqqehcW1SmPBF1MfXgk3m1MukeAKILuABg="},"BlockNumber":8415189}} <nil>
{"MessageType":5,"Point":{"Slot":85162758,"Hash":"6kV9P2zOMOqqehcW1SmPBF1MfXgk3m1MukeAKILuABg="},"Tip":{"Point":{"Slot":85162758,"Hash":"6kV9P2zOMOqqehcW1SmPBF1MfXgk3m1MukeAKILuABg="},"BlockNumber":8415189}} <nil>
{"MessageType":3,"Point":{"Slot":85162758,"Hash":"6kV9P2zOMOqqehcW1SmPBF1MfXgk3m1MukeAKILuABg="},"Tip":{"Point":{"Slot":85162758,"Hash":"6kV9P2zOMOqqehcW1SmPBF1MfXgk3m1MukeAKILuABg="},"BlockNumber":8415189}} <nil>
roll backward: point = common.Point{_:struct {}{}, Slot:0x5137b06, Hash:[]uint8{0xea, 0x45, 0x7d, 0x3f, 0x6c, 0xce, 0x30, 0xea, 0xaa, 0x7a, 0x17, 0x16, 0xd5, 0x29, 0x8f, 0x4, 0x5d, 0x4c, 0x7d, 0x78, 0x24, 0xde, 0x6d, 0x4c, 0xba, 0x47, 0x80, 0x28, 0x82, 0xee, 0x0, 0x18}}, tip = chainsync.Tip{_:struct {}{}, Point:common.Point{_:struct {}{}, Slot:0x5137b06, Hash:[]uint8{0xea, 0x45, 0x7d, 0x3f, 0x6c, 0xce, 0x30, 0xea, 0xaa, 0x7a, 0x17, 0x16, 0xd5, 0x29, 0x8f, 0x4, 0x5d, 0x4c, 0x7d, 0x78, 0x24, 0xde, 0x6d, 0x4c, 0xba, 0x47, 0x80, 0x28, 0x82, 0xee, 0x0, 0x18}}, BlockNumber:0x8067d5}
{"MessageType":1} <nil>
{"MessageType":2,"WrappedHeader":{"Era":5,"RawMessage":null},"Tip":{"Point":{"Slot":85162758,"Hash":"6kV9P2zOMOqqehcW1SmPBF1MfXgk3m1MukeAKILuABg="},"BlockNumber":8415189}} <nil>
{"MessageType":1} <nil>
era = Babbage, slot = 85162763, block_no = 8415190, id = 2be27a0c8f7df130d2c876e9f696df9eac714daf1108f6e0474a9fe48ca766d6
{"MessageType":2,"WrappedHeader":{"Era":5,"RawMessage":null},"Tip":{"Point":{"Slot":85162763,"Hash":"K+J6DI998TDSyHbp9pbfnqxxTa8RCPbgR0qf5IynZtY="},"BlockNumber":8415190}} <nil>
{"MessageType":1} <nil>
era = Babbage, slot = 85162766, block_no = 8415191, id = 0c8e5165f83e3128efc97c0179408c71a35d7d8fadd5e8c8e4a587ee35b7923b
{"MessageType":2,"WrappedHeader":{"Era":5,"RawMessage":null},"Tip":{"Point":{"Slot":85162766,"Hash":"DI5RZfg+MSjvyXwBeUCMcaNdfY+t1ejI5KWH7jW3kjs="},"BlockNumber":8415191}} <nil>
{"MessageType":1} <nil>
ERROR: protocol error: decode error: cbor: invalid map key type: []uint8

**Comment by zhp1254** [comment]
> git status HEAD detached at v0.29.0 Untracked files: (use "git add ..." to include in what will be committed)
> 
> ```
> cmd/go-ouroboros-network/go-ouroboros-network
> ```
> 
> ./go-ouroboros-network -network mainnet -socket /cardano-node/state-node-mainnet/node.socket chain-sync gt roll backward: point = common.Point{_:struct {}{}, Slot:0x0, Hash:[]uint8(nil)}, tip = chainsync.Tip{_:struct {}{}, Point:common.Point{_:struct {}{}, Slot:0x51372f1, Hash:[]uint8{0x7b, 0xb3, 0x2d, 0x98, 0x6d, 0x64, 0x85, 0xb2, 0x14, 0xe2, 0x58, 0x2f, 0x6c, 0xb7, 0x4c, 0xf9, 0x3c, 0x8f, 0x17, 0xfd, 0x65, 0xba, 0xc1, 0x12, 0x52, 0xbc, 0xe9, 0x38, 0xcd, 0xd2, 0xa3, 0xfd}}, BlockNumber:0x806786} ERROR: protocol error: decode error: EOF
> 
> when i Println in protocol/chainsync/client.go line :212 i got an error: MsgRollForwardNtC ledger.NewBlockFromCbor: 0 [] ![image](https://user-images.githubusercontent.com/18046237/219869555-2abda570-fb90-42fd-bcbf-1a0cf894b890.png)

fixed with code:
```
msg := msgGeneric.(*MsgRollForwardNtC)
		var wrappedBlock WrappedBlock

		switch msg.WrappedBlock.Content.(type) {
		case string:
			wrappedBlockContent, _ := base64.StdEncoding.DecodeString(msg.WrappedBlock.Content.(string))
			utils.CborDecode(wrappedBlockContent, &wrappedBlock)
		case  []byte:
			utils.CborDecode(msg.WrappedBlock.Content.([]byte), &wrappedBlock)
		default:
			fmt.Println("MsgRollForwardNtC unknown type")
		}

		blk, err := ledger.NewBlockFromCbor(wrappedBlock.BlockType, wrappedBlock.BlockCbor)
		if err != nil {
			return err
		}
		// Call the user callback function
		return c.config.RollForwardFunc(wrappedBlock.BlockType, blk, msg.Tip)
```

**Comment by zhp1254** [comment]
./go-ouroboros-network  -network mainnet -socket  state-node-mainnet/node.socket chain-sync -start-era alonzo


era = Alonzo, slot = 43103186, block_no = 6392068, id = ab16da8c5ee420fbf8878791d70acdb3b13e05e1bec514449a20fc65de9867e8 trans = 58
era = Alonzo, slot = 43103220, block_no = 6392069, id = a1531fa055ef8722a2a9e1c545521a54b0fb8d945cca2d6c0290bca2f8b1c09c trans = 35
era = Alonzo, slot = 43103226, block_no = 6392070, id = 8443795ea9a403dde5ee71480b1a68e7daa2fac62c7f4a645407432fc9c9a49d trans = 6
panic: runtime error: hash of unhashable type []interface {}

goroutine 41 [running]:
reflect.mapassign(0x7122a0, 0xc0004a3160?, 0x0?, 0x774356?)
	/data/go/src/runtime/map.go:1353 +0x25
reflect.Value.SetMapIndex({0x7122a0?, 0xc000014770?, 0xc0004a3160?}, {0x7229a0, 0xc0004a3110, 0x199}, {0x7229a0, 0xc0004a3160, 0x199})
	/data/go/src/reflect/value.go:2325 +0x391
github.com/fxamacker/cbor/v2.(*decoder).parseMapToMap(0xc0000a2f58, {0x7122a0?, 0xc000014770?, 0x7020c0?}, 0xc00030e870)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1295 +0xac7
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a2f58, {0x7122a0?, 0xc000014770?, 0xdf?}, 0xc0000a2ee8?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:782 +0xef8
github.com/fxamacker/cbor/v2.(*decoder).value(0xc0000a2f58, {0x6f8aa0?, 0xc000014770?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:509 +0x325
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc0000a2f48, {0x6f8aa0, 0xc000014770})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:36 +0xb1
github.com/cloudstruct/go-cardano-ledger/cbor.Decode({0xc000505402, 0xdf, 0x1fe}, {0x6f8aa0, 0xc000014770})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/decode.go:12 +0xee
github.com/cloudstruct/go-cardano-ledger/cbor.(*Value).UnmarshalCBOR(0xc000304ab0, {0xc000505402, 0xdf, 0x1fe})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/value.go:14 +0x2e5
github.com/fxamacker/cbor/v2.(*decoder).parseToUnmarshaler(0xc0000a3908, {0x7229a0?, 0xc000304ab0?, 0x418953?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:891 +0x162
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a3908, {0x7229a0?, 0xc000304ab0?, 0x2?}, 0xc00046f9c8?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:644 +0x449
github.com/fxamacker/cbor/v2.(*decoder).parseArrayToSlice(0xc0000a3908, {0x6fd520?, 0xc00046f9b0?, 0x7020c0?}, 0xc00030e690)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1124 +0x28c
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a3908, {0x6fd520?, 0xc00046f9b0?, 0xe1?}, 0xc0000a3898?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:770 +0xde6
github.com/fxamacker/cbor/v2.(*decoder).value(0xc0000a3908, {0x6f4820?, 0xc00046f9b0?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:509 +0x325
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc0000a38f8, {0x6f4820, 0xc00046f9b0})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:36 +0xb1
github.com/cloudstruct/go-cardano-ledger/cbor.Decode({0xc000505002, 0xe1, 0x1fe}, {0x6f4820, 0xc00046f9b0})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/decode.go:12 +0xee
github.com/cloudstruct/go-cardano-ledger/cbor.(*Value).UnmarshalCBOR(0xc000304a90, {0xc000505002, 0xe1, 0x1fe})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/value.go:25 +0x392
github.com/fxamacker/cbor/v2.(*decoder).parseToUnmarshaler(0xc0000a42b8, {0x7229a0?, 0xc000304a90?, 0xc0004a3050?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:891 +0x162
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a42b8, {0x7229a0?, 0xc000304a90?, 0x2?}, 0xc00046f998?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:644 +0x449
github.com/fxamacker/cbor/v2.(*decoder).parseArrayToSlice(0xc0000a42b8, {0x6fd520?, 0xc00046f980?, 0xc0004db048?}, 0xc00030e690)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1124 +0x28c
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a42b8, {0x6fd520?, 0xc00046f980?, 0xe3?}, 0xc0000a4248?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:770 +0xde6
github.com/fxamacker/cbor/v2.(*decoder).value(0xc0000a42b8, {0x6f4820?, 0xc00046f980?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:509 +0x325
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc0000a42a8, {0x6f4820, 0xc00046f980})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:36 +0xb1
github.com/cloudstruct/go-cardano-ledger/cbor.Decode({0xc000136d34, 0xe3, 0x70cc}, {0x6f4820, 0xc00046f980})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/decode.go:12 +0xee
github.com/cloudstruct/go-cardano-ledger/cbor.(*Value).UnmarshalCBOR(0xc0004a30e0, {0xc000136d34, 0xe3, 0x70cc})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/value.go:25 +0x392
github.com/fxamacker/cbor/v2.(*decoder).parseToUnmarshaler(0xc0000a5ad8, {0x7229a0?, 0xc0004a30e0?, 0x7122a0?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:891 +0x162
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a5ad8, {0x7229a0?, 0xc0004a30e0?, 0x1?}, 0xc00046f968?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:644 +0x449
github.com/fxamacker/cbor/v2.(*decoder).parseArrayToSlice(0xc0000a5ad8, {0x6fd520?, 0xc000144288?, 0xc000144280?}, 0xc00030e690)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1124 +0x28c
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a5ad8, {0x6fd520?, 0xc000144288?, 0xc00031e1c0?}, 0x4?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:770 +0xde6
github.com/fxamacker/cbor/v2.(*decoder).parseMapToStruct(0xc0000a5ad8, {0x74b220?, 0xc000144248?, 0xdfee001a3905220d?}, 0xc00030e320)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1586 +0xdd3
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc0000a5ad8, {0x74b220?, 0xc000144248?, 0x58?}, 0xc0005a2198?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:780 +0xf0d
github.com/fxamacker/cbor/v2.(*decoder).parseArrayToSlice(0xc0000a5ad8, {0x6fd1a0?, 0xc000300f08?, 0x729060?}, 0xc00030e2d0)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1124 +0x28c
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc00054fad8, {0x6fd1a0?, 0xc000300f08?, 0x2?}, 0x501757?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:770 +0xde6
github.com/fxamacker/cbor/v2.(*decoder).parseArrayToStruct(0xc0000a5ad8, {0x74eec0?, 0xc000300f00?, 0x2?}, 0xc00030e230)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:1379 +0x4ac
github.com/fxamacker/cbor/v2.(*decoder).parseToValue(0xc00054fad8, {0x74eec0?, 0xc000300f00?, 0xc00012e000?}, 0x0?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:774 +0xdbc
github.com/fxamacker/cbor/v2.(*decoder).value(0xc00054fad8, {0x733a40?, 0xc000300f00?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/decode.go:509 +0x325
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc0000a5ac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:36 +0xb1
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc00054fac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:47 +0x135
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc00054fac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:47 +0x135
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc00054fac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:47 +0x135
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc00054fac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:47 +0x135
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc00054fac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:47 +0x135
github.com/fxamacker/cbor/v2.(*Decoder).Decode(0xc00054fac8, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/fxamacker/cbor/v2/stream.go:47 +0x135
github.com/cloudstruct/go-cardano-ledger/cbor.Decode({0xc00011e000, 0xf33a, 0xf33a}, {0x733a40, 0xc000300f00})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/cbor/decode.go:12 +0xee
github.com/cloudstruct/go-cardano-ledger.NewAlonzoBlockFromCbor({0xc00011e000, 0xf33a, 0xf33a})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/alonzo.go:81 +0x56
github.com/cloudstruct/go-cardano-ledger.NewBlockFromCbor(0xc0000cc000?, {0xc00011e000?, 0xf33c?, 0x6f7020?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/vendor/github.com/cloudstruct/go-cardano-ledger/block.go:47 +0x19b
github.com/cloudstruct/go-ouroboros-network/protocol/chainsync.(*Client).handleRollForward(0xc0002a4040, {0x7fe898?, 0xc0002aa300?})
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/protocol/chainsync/client.go:222 +0x4d3
github.com/cloudstruct/go-ouroboros-network/protocol/chainsync.(*Client).messageHandler(0xc0002a4040, {0x7fe898, 0xc0002aa300}, 0xd0?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/protocol/chainsync/client.go:86 +0x65
github.com/cloudstruct/go-ouroboros-network/protocol.(*Protocol).handleMessage(0xc0002b0000, {0x7fe898, 0xc0002aa300}, 0x20?)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/protocol/protocol.go:410 +0x16b
github.com/cloudstruct/go-ouroboros-network/protocol.(*Protocol).recvLoop(0xc0002b0000)
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/protocol/protocol.go:322 +0x46e
created by github.com/cloudstruct/go-ouroboros-network/protocol.(*Protocol).Start
	/data/gopath/src/github/cloudstruct/go-ouroboros-network/protocol/protocol.go:115 +0x238

**Comment by agaffney** [comment]
> fixed with code:
> 
> ```
> msg := msgGeneric.(*MsgRollForwardNtC)
> 		var wrappedBlock WrappedBlock
> 
> 		switch msg.WrappedBlock.Content.(type) {
> 		case string:
> 			wrappedBlockContent, _ := base64.StdEncoding.DecodeString(msg.WrappedBlock.Content.(string))
> 			utils.CborDecode(wrappedBlockContent, &wrappedBlock)
> 		case  []byte:
> 			utils.CborDecode(msg.WrappedBlock.Content.([]byte), &wrappedBlock)
> 		default:
> 			fmt.Println("MsgRollForwardNtC unknown type")
> 		}
> 
> 		blk, err := ledger.NewBlockFromCbor(wrappedBlock.BlockType, wrappedBlock.BlockCbor)
> 		if err != nil {
> 			return err
> 		}
> 		// Call the user callback function
> 		return c.config.RollForwardFunc(wrappedBlock.BlockType, blk, msg.Tip)
> ```

Are you saying that some `MsgRollForwardNtC` will have the block data CBOR additionally base64-encoded? Did this change in some recent era?

**Comment by agaffney** [comment]
FYI, I used a recent version of the test program from this library to run through the entire `mainnet` chain via chain-sync and parse each block, without any issue

**Comment by agaffney** [comment]
This appears to be the same problem as #232, which was fixed by #233, so I'm closing this
